app/models/wg_config.rb

```diff
@@ -0,0 +1,81 @@
+require 'ipaddress'
+require 'open3'
+
+class WgConfig < ApplicationRecord
+
+  belongs_to :user
+
+  validates :name, presence: true
+  validates :privatekey, presence: true
+  validates :publickey, presence: true
+  validates :name, uniqueness: { scope: :user_id }
+
+  after_initialize :init_keys
+
+  def generate_keys!
+    self.privatekey = gen_privatekey
+    self.publickey = gen_publickey(self.privatekey)
+  end
+
+  def addresses(as_networks=true)
+    raise "Model is not saved" if self.id.nil?
+
+    first = IPAddress Setting['wgFirstClientAddress']
+    last = IPAddress Setting['wgLastClientAddress']
+
+    ip = IPAddress(first.to_i + self.id)
+
+    raise "Failed to assign IP: out of range" if ip > last
+
+    ip.netmask = Setting['wgNetmask'] if as_networks
+
+    ip.to_string
+  end
+
+  def to_s
+    lines = []
+    lines << "# WireGuard config #{name}"
+    lines << '[Interface]'
+    lines << "PrivateKey = #{privatekey}"
+    lines << "Address = #{self.addresses}"
+    lines << ''
+    lines << '[Peer]'
+    lines << "PublicKey = #{Setting['wgServerPublicKey']}"
+    lines << "Endpoint = #{Setting['wgServerEndpoint']}"
+    lines << "AllowedIPs = #{Setting['wgAllowedIPs']}"
+
+    lines.join("\n")
+  end
+
+  def as_peer
+    s = []
+    s << "# User #{self.user_id}, config '#{self.name}'"
+    s << "[Peer]"
+    s << "PublicKey = #{publickey}"
+    s << "AllowedIPs = #{addresses(false)}"
+
+    s.join("\n")
+  end
+
+  private
+
+  def init_keys
+    generate_keys! if self.privatekey.blank?
+  end
+
+  def gen_privatekey
+    key, status = Open3.capture2("wg genkey")
+    key.strip
+  end
```

Comment:
If we had to call external programs here, I would just wrap the whole thing in shell script, so that it could run independently of Ruby.